File: internal/interpreters/python_requires.go
package interpreters

// Copyright (C) 2025 by Posit Software, PBC.

import (
	"errors"
	"fmt"
	"regexp"
	"strings"

	"github.com/posit-dev/publisher/internal/util"

	"gopkg.in/ini.v1"

	toml "github.com/pelletier/go-toml/v2"
)

type PyProjectPythonRequires struct {
	ProjectPath util.AbsolutePath
}

func NewPyProjectPythonRequires(projectPath util.AbsolutePath) *PyProjectPythonRequires {
	return &PyProjectPythonRequires{
		ProjectPath: projectPath,
	}
}

// Find the python version requested by the project if specified in any of the
// supported metadata files. The order of precedence is:
// 1. .python-version
// 2. pyproject.toml
// 3. setup.cfg
//
// The version specifications are the one defined by PEP 440
// if no version is found, an error is returned.
func (p *PyProjectPythonRequires) GetPythonVersionRequirement() (string, error) {
	if version, err := p.readPythonVersionFile(); err == nil && version != "" {
		return version, nil
	}
	if version, err := p.readPyProjectToml(); err == nil && version != "" {
		return version, nil
	}
	if version, err := p.readSetupCfg(); err == nil && version != "" {
		return version, nil
	}
	return "", errors.New("no python version requirement found")
}

// Read a .python-version file and return the version string.
// the file is a plain text file that contains only the version specification.
func (p *PyProjectPythonRequires) readPythonVersionFile() (string, error) {
	path := p.ProjectPath.Join(".python-version")
	data, err := path.ReadFile()
	if err != nil {
		return "", err
	}

	parts := strings.Split(string(data), ",")
	var adapted []string

	for _, part := range parts {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		adaptedPart, err := adaptPythonRequires(part)
		if err != nil {
			return "", err
		}
		adapted = append(adapted, adaptedPart)
	}

	return strings.Join(adapted, ","), nil
}

// Read a pyproject.toml file and return the version string.
// The file is a TOML file that contains a [project] section
// with a requires-python key.
//
// [project]
// requires-python = ">=3.8"
func (p *PyProjectPythonRequires) readPyProjectToml() (string, error) {
	path := p.ProjectPath.Join("pyproject.toml")
	data, err := path.ReadFile()
	if err != nil {
		return "", err
	}

	var tomlData map[string]any
	if err := toml.Unmarshal(data, &tomlData); err != nil {
		return "", err
	}

	// project.requires-python
	if project, ok := tomlData["project"].(map[string]any); ok {
		if req, ok := project["requires-python"]; ok {
			return fmt.Sprintf("%v", req), nil
		}
	}

	return "", nil
}

// Read a setup.cfg file and return the version string.
// The file is an INI file that contains an [options] section
// with a python_requires key.
//
// [options]
// python_requires = ">=3.8"
func (p *PyProjectPythonRequires) readSetupCfg() (string, error) {
	path := p.ProjectPath.Join("setup.cfg")
	if exists, _ := path.Exists(); !exists {
		return "", errors.New("setup.cfg file does not exist")
	}

	cfg, err := ini.Load(path.String())
	if err != nil {
		return "", err
	}

	// options.python_requires
	if section, err := cfg.GetSection("options"); err == nil {
		if key, err := section.GetKey("python_requires"); err == nil {
			return key.String(), nil
		}
	}

	return "", nil
}

var pep440Operators = regexp.MustCompile(`(==|!=|<=|>=|~=|<|>)`)
var validVersion = regexp.MustCompile(`^\d+(\.\d+)*(\.\*)?$`)

func adaptPythonRequires(raw string) (string, error) {
	constraint := strings.TrimSpace(raw)

	if strings.ContainsAny(constraint, "-/@") {
		return "", fmt.Errorf("python specific implementations are not supported: %s", constraint)
	}
	if strings.Contains(constraint, "rc") || strings.Contains(constraint, "b") || strings.Contains(constraint, "a") {
		return "", fmt.Errorf("pre-release versions are not supported: %s", constraint)
	}

	if strings.Count(constraint, ".") > 2 {
		return "", fmt.Errorf("invalid python version: %s", constraint)
	}

	// If it's already a PEP 440 constraint, return it as is
	if pep440Operators.MatchString(constraint) {
		return constraint, nil
	}

	// Otherwise it should be a version string
	if !validVersion.MatchString(constraint) {
		return "", fmt.Errorf("invalid python version: %s", constraint)
	}

	// If the version doesn't have an operator,
	// but has a dot, we use equivalence.
	// e.g. 3.8.* -> ==3.8.*
	if strings.Contains(constraint, "*") {
		return "==" + constraint, nil
	}

	// strip trailing zeros and dots
	// e.g. 3.8.0 -> 3.8
	// and prefix with ~=
	trimmed := strings.TrimSuffix(constraint, ".0")
	for strings.HasSuffix(trimmed, ".0") {
		trimmed = strings.TrimSuffix(trimmed, ".0")
	}
	return "~=" + trimmed, nil
}

File: extensions/vscode/src/interpreters/pythonRequires.ts
// Copyright (C) 2026 by Posit Software, PBC.

import path from "node:path";
import { readFileText } from "./fsUtils";
import { adaptPythonRequires } from "./versionConstraints";

/**
 * Find the Python version requested by the project, checking in order:
 * 1. .python-version
 * 2. pyproject.toml
 * 3. setup.cfg
 *
 * Returns the PEP 440 version specification, or empty string if not found.
 */
export async function getPythonRequires(projectDir: string): Promise<string> {
  const fromVersionFile = await readPythonVersionFile(projectDir);
  if (fromVersionFile) {
    return fromVersionFile;
  }

  const fromPyProject = await readPyProjectToml(projectDir);
  if (fromPyProject) {
    return fromPyProject;
  }

  const fromSetupCfg = await readSetupCfg(projectDir);
  if (fromSetupCfg) {
    return fromSetupCfg;
  }

  return "";
}

/**
 * Read .python-version file. Plain text, possibly comma-separated versions.
 * Each part is adapted through adaptPythonRequires().
 */
async function readPythonVersionFile(
  projectDir: string,
): Promise<string | undefined> {
  const content = await readFileText(path.join(projectDir, ".python-version"));
  if (content === null) {
    return undefined;
  }

  const parts = content.split(",");
  const adapted: string[] = [];

  for (const part of parts) {
    const trimmed = part.trim();
    if (!trimmed) {
      continue;
    }
    const result = adaptPythonRequires(trimmed);
    if (result === null) {
      return "";
    }
    adapted.push(result);
  }

  return adapted.join(",") || undefined;
}

const requiresPythonRe = /^\s*requires-python\s*=\s*["']([^"']+)["']\s*$/m;

/**
 * Read pyproject.toml and extract requires-python from [project] section.
 * Uses regex extraction instead of a full TOML parser.
 */
async function readPyProjectToml(
  projectDir: string,
): Promise<string | undefined> {
  const content = await readFileText(path.join(projectDir, "pyproject.toml"));
  if (content === null) {
    return undefined;
  }

  const match = requiresPythonRe.exec(content);
  if (match && match[1]) {
    return match[1];
  }

  return undefined;
}

/**
 * Read setup.cfg and extract python_requires from [options] section.
 * Simple line-by-line INI parsing.
 */
async function readSetupCfg(projectDir: string): Promise<string | undefined> {
  const content = await readFileText(path.join(projectDir, "setup.cfg"));
  if (content === null) {
    return undefined;
  }

  const lines = content.split("\n");
  let inOptionsSection = false;

  for (const line of lines) {
    const trimmed = line.trim();

    // Section header
    if (trimmed.startsWith("[")) {
      inOptionsSection = trimmed.toLowerCase() === "[options]";
      continue;
    }

    if (inOptionsSection) {
      // Match key = value within the [options] section
      const eqIdx = trimmed.indexOf("=");
      if (eqIdx !== -1) {
        const key = trimmed.substring(0, eqIdx).trim();
        if (key === "python_requires") {
          return trimmed.substring(eqIdx + 1).trim();
        }
      }
    }
  }

  return undefined;
}
